internal/unix/types_linux.go

```diff
@@ -202,22 +205,44 @@ func KernelRelease() (string, error) {
 }
 
 func RemoveMemlockRlimit() (func() error, error) {
-	oldLimit := new(Rlimit)
-	if err := linux.Getrlimit(RLIMIT_MEMLOCK, oldLimit); err != nil {
-		return nil, fmt.Errorf("failed to get memlock rlimit: %w", err)
-	}
-
-	if err := linux.Setrlimit(RLIMIT_MEMLOCK, &Rlimit{
+	ol, err := prlimit(RLIMIT_MEMLOCK, Rlimit{
 		Cur: RLIM_INFINITY,
 		Max: RLIM_INFINITY,
-	}); err != nil {
+	})
+	if err != nil {
 		return nil, fmt.Errorf("failed to set memlock rlimit: %w", err)
 	}
 
 	return func() error {
-		if err := linux.Setrlimit(RLIMIT_MEMLOCK, oldLimit); err != nil {
-			return fmt.Errorf("failed to reset memlock rlimit: %w", err)
+		if err := linux.Setrlimit(RLIMIT_MEMLOCK, ol); err != nil {
+			return fmt.Errorf("failed to revert memlock rlimit: %w", err)
 		}
 		return nil
 	}, nil
 }
+
+// prlimit wraps the prlimit64 syscall. It is used for getting and setting
+// rlimit values atomically. Setting the pid is not supported, as that's
+// outside of the scope of the library.
+func prlimit(resource int, newLimit Rlimit) (*Rlimit, error) {
+	var oldLimit Rlimit
+
+	ol := unsafe.Pointer(&oldLimit)
+	nl := unsafe.Pointer(&newLimit)
+
+	_, _, errno := linux.RawSyscall6(SYS_PRLIMIT64,
+		0, // pid 0 means own process
+		uintptr(resource),
+		uintptr(nl),
+		uintptr(ol),
+		0, 0)
+
+	runtime.KeepAlive(ol)
```

Comment:
Do the `uintptr(unsafe.Pointer)` conversion and you don't need these, see point (4) https://pkg.go.dev/unsafe#Pointer

Reply:
Ah, thanks! Was wondering if that was valid. :+1: